source-oracle/source_oracle/resources.py

```diff
@@ -0,0 +1,151 @@
+                msg = f"We require a minimum of 7 days UNDO_RETENTION to ensure consistency of this task. The current UNDO_RETENTION is {undo_retention_seconds} seconds. See go.estuary.dev/source-oracle for more information on how to configure the UNDO_RETENTION."  # nopep8
+                if skip_retention_checks:
+                    log.warn(msg)
+                else:
+                    raise Exception(msg)
+
+            if avg_retention_seconds < one_week_seconds:
+                msg = f"We require a minimum of 7 days undo retention to ensure consistency of this task. The current average auto-tuned retention of the database for the past four days is {avg_retention_seconds} seconds. See go.estuary.dev/source-oracle for more information on how to configure the UNDO_RETENTION."  # nopep8
+                if skip_retention_checks:
+                    log.warn(msg)
+                else:
+                    raise Exception(msg)
+
+            if not autoextensible:
+                log.warn("We recommend making your undo tablespace auto-extensible. See go.estuary.dev/source-oracle for more information.")
+
+            if retention_mode != 'GUARANTEE':
+                log.warn("We recommend guaranteeing retention of the undo tablespace. See go.estuary.dev/source-oracle for more information.")
+
+
+async def all_resources(
+    log: Logger, http: HTTPMixin, config: EndpointConfig, pool: oracledb.AsyncConnectionPool,
+) -> list[common.Resource]:
+    resources_list = []
+
+    oracle_tables = await fetch_tables(log, pool)
+    oracle_columns = await fetch_columns(log, pool)
+
+    current_scn = None
+    async with pool.acquire() as conn:
+        with conn.cursor() as c:
+            await c.execute("SELECT current_scn FROM V$DATABASE")
+            current_scn = (await c.fetchone())[0]
+
+    for ot in oracle_tables:
+        columns = [col for col in oracle_columns if col.table_name == ot.table_name]
+        t = build_table(config, ot.table_name, columns)
+
+        max_rowid = None
+        async with pool.acquire() as conn:
+            with conn.cursor() as c:
+                await c.execute(f"SELECT max(ROWID) FROM {t.table_name}")
+                max_rowid = (await c.fetchone())[0]
+        # if max_rowid is None, that maens there are no rows in the table, so we
+        # skip backfill
+        backfill = ResourceState.Backfill(cutoff=(max_rowid,)) if max_rowid is not None else None
+
+        def open(
+            table: Table,
+            binding: CaptureBinding[ResourceConfig],
+            binding_index: int,
+            state: ResourceState,
+            task: Task,
+            _all_bindings,
+        ):
+            common.open_binding(
+                binding,
+                binding_index,
+                state,
+                task,
+                fetch_page=functools.partial(fetch_page, table, pool),
+                fetch_changes=functools.partial(fetch_changes, table, pool),
```

Comment:
Yep, they run concurrently, and the coordination between them is through the `cutoff` parameter of backfill which is supposed to separate the documents the backfill and incremental changes are captured.

The initial state of backfill has a cutoff set to the last `ROWID` of the table we operate on, and the starting point of the incremental capture is set to the current SCN of the database.

Thinking about this now, I wonder if there is still risk of race between them. I think there is a chance that while we are processing the documents from the backfill query, some document changes, the incremental task outputs the change, but we later process the same document in the backfill task and emit an older version of it 🤔

Reply:
Yeah. This is the exact problem that the whole watermarking approach we use in the other SQL CDC connectors is intended to solve. We assume that if we write a watermark, read some rows from a table and then write a second watermark, then by the time we observe the watermark via replication we must necessarily also have observed all changes which resulted in the backfilled state of that portion of the table, plus zero or more changes beyond that. And by emitting the backfilled rows immediately but hold off on emitting a checkpoint update until we observe that next watermark change (only when there's a backfill chunk like this, other replication streaming doesn't do this because that could result in massive transactions) we guarantee that the observable state is both correct and respects transaction boundaries.

There may be some heuristic you can use to simplify things here, but in general if you don't trust there to be any particular causal relationship between the stream of changes and the backfilled row states you need to do something to establish that causality.

Reply:
@willdonnelly so am I right that we can handle this by having the state of the backfill task available to fetch_changes, and filtering events of fetch_changes which have `ROWID`s that have been processed by the backfill task based on the state. So whenever the backfill task emits N rows, it also updates its last processed ROWID, and the fetch_changes task will only process events which have a ROWID below the backfilled one. If the ROWID is higher, we pause the fetch_changes task (well, we loop in it until the last ROWID has progressed). Once the backfill task has finished, we then don't need to filter anymore in fetch_changes.

Implementing this requires some tweaks in the estuary_cdk 🤔

Reply:
Hmm, I'm having trouble thinking this through right now because it's always a really complicated subject to get right, but reasoning by analogy the ROWID comparison you propose seems like it would serve the same purpose as the primary-key comparisons we do in the precise backfill mode. And the reason we have that is just to ensure that the first time a document is observed in the capture's output, it shows up as an insertion (whether from replication or a "pretend" insert event from the backfill) and not an update/deletion of the row. The watermarking is a different piece of the puzzle, so I suspect that the ROWID comparison alone wouldn't do the trick anyway.

Reply:
Ah, I just read your comment a bit more closely and I think there's another more fundamental issue with the approach you describe. The bit where "if the ROWID is higher, we wait until the ROWID is increased by the backfill job then we emit the changes" would effectively mean that a new row inserted at the end of the table would block change streaming until the backfill completed.

And of course "do the whole backfill and only after it completes begin streaming changes from a point in time prior to the backfill start" _is_ a reliable policy which ensures a consistent ordering of all change events, so that suggests that your proposed logic would actually be correct. It's just also not super helpful since most tables will have writes that either append to the end or are uniformly distributed throughout the table, and either of those ends up approximating the lazy "don't even start streaming changes until we're done" policy so closely that we might as well just do that instead and avoid the extra complexity.

Reply:
Thinking this through step by step:

- The primary concern is about eventual consistency. It's nice to make guarantees about intermediate states where possible, but the really important thing is just that if the user doesn't change a particular row and our capture is allowed to get caught up, the row will eventually have the correct state in the collection.
- If we consider change streaming in isolation without any filtering, it's trivially correct in this sense for any row modified after the capture started. The issue is how to bring in backfilled row states in a way that doesn't compromise correctness.
- If we just blindly run the two in parallel, the concern is that if a row is changed concurrently with our backfill of that portion of the table, it's possible for the backfill query to observe the old value but the change event get emitted before the backfilled version, which means that the row is now in an incorrect state which will persist unless/until it gets modified again in the source DB.
- Blocking change streaming until we've backfilled up to a given row ID is a valid solution to avoid this, but in the worst case it reduces to simply waiting for the backfill to complete before processing any changes, and almost any concurrent-updates scenario will closely approximate that worst-case.
- Filtering out change events with a ROWID greater than the last backfilled one works to avoid the backfill-after-replication concurrency issue, but introduces a new failure mode where we could theoretically filter out change events which ought to have been emitted.
- This whole concurrency issue is what we use watermarks to deal with. Once we see the next watermark we know we have definitely processed any changes which could have occurred concurrently with our backfill query, and so when we emit the backfilled rows followed by all changes since the previous watermark we guarantee the "changes after backfill" ordering invariant. I've just realized that this is in some sense very similar to a "backfill the whole table, then stream all changes since the start of the backfill" policy just carried out on a much smaller scale.
- The thing that I keep coming back to is that in this Oracle capture the backfill and change stream threads aren't completely uncoordinated. They're both executing queries synchronously against the database, and it feels like it might be possible to use that instead of watermarks to get the same sort of "backfill then changes" ordering.
- Hand-waving a bunch here, but suppose that we removed the watermarking behavior from other SQL CDC captures and instead replaced it with a "stream changes for 5 minutes" policy. So long as 5 minutes is enough for us to process all outstanding change events and get caught up to "now" _and_ for the database to not do any weird stale-read concurrency stuff in practice, this would work just as well as actually writing and waiting for a watermark.
- And it doesn't actually have to be 5 minutes, that's just an example of the dumbest possible solution that would work in most cases. More likely what we'd do is some sort of heuristic which could figure out when we're basically caught up on change streaming by querying the system max LSN and streaming to there, plus another second or two at most.
- But something like that would still hinge on some sort of blocking coordination between the backfill and change streaming processes, which would iteratively backfill a chunk and then stream changes until fully caught up to "now" on the source database.

Reply:
I think there are still only three options here that actually preserve correctness in all possible edge cases:

- The laziest possible option: just run the entire backfill and only then begin streaming all changes since the backfill first began. Since we're requiring 7 days of change retention anyway that's a fairly generous amount of time to finish the backfill.
- Use a watermarks table just like other SQL CDC captures, in order to establish whether we've become caught up to "now" while processing replication events. It's heavyweight and complex, but there's a **_reason_** that's the design we arrived at when we first sat down to figure out how to do accurate SQL CDC without a full snapshot or a backfill-everything-then-stream arrangement.
- Essentially the same logic as watermarking, but using some sort of "get current max SCN of the database and stream to that" instead of actual watermark table writes. This would be breaking new ground and requires the same exact pattern of coordination inside the connector, but would avoid the requirement for a writable watermarks table in the Oracle DB.

Reply:
Here is a sketch of what might work:
1. We start a backfill and a changes task concurrently
2. Since we have identified the table requires backfilling, the changes task will not proceed since the rowid of any changes captured will be larger than the starting rowid of `AAAAAAA...`
3. So we are guaranteed that the first task to run is a backfill task. Meanwhile runs of the changes task will not proceed, since there is still no checkpoint from backfill task. A backfill task runs and emits 1000 documents and a checkpoint, along with an integer counter which is used to synchronise backfill and changes tasks.
4. At this point the changes task will fetch any changes from the time of step (1) to the last transaction of the database (`BETWEEN {cursor} and MAXVALUE`). While this is happening, the backfill task will not run due to the counter being an even number (we count 0 as even). Once the incremental task is done, we increment the counter by one, at which point it becomes odd and the backfill task is then allowed to run.
5. This ensures a ping-pong between backfill and changes tasks, so we backfill up to a certain point, then fetch all changes since the last time we fetched changes until the last transaction of the database